libcxx/include/__numeric/saturation_arithmetic.h

```diff
@@ -0,0 +1,119 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___NUMERIC_SATURATION_ARITHMETIC_H
+#define _LIBCPP___NUMERIC_SATURATION_ARITHMETIC_H
+
+#include <__concepts/arithmetic.h>
+#include <__config>
+#include <__type_traits/decay.h>
+#include <__utility/cmp.h>
+#include <limits>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#if _LIBCPP_STD_VER >= 26
+
+template <typename _Tp>
+concept __libcpp_standard_integer = __libcpp_unsigned_integer<decay_t<_Tp>> || __libcpp_signed_integer<decay_t<_Tp>>;
+
+template <__libcpp_standard_integer _Tp>
+_LIBCPP_HIDE_FROM_ABI constexpr _Tp add_sat(_Tp __x, _Tp __y) noexcept {
+  if (_Tp __sum; !__builtin_add_overflow(__x, __y, &__sum))
```

Comment:
Couldn't we just extend `__builtin_add_overflow` to accept extended integer types?

Reply:
How does that differ from what I asked Aaron?

Reply:
> These builtins don't work with 128-bit integrals. @AaronBallman are you happy to receive patches for this in Clang? If so what is the proper name for `__builtin_uaddXX_overflow` and `____builtin_saddXX_overflow`? Obviously there should be patches for `sub` and `mul`.
> 
> @H-G-Hristov if acceptable for Clang do you want to do the work?

Yes, I could try to do the work. Please note I have no experience with Clang yet.

Reply:
@mordante It sounded to me like you've asked for additional builtins for extended integer types instead of extending the existing one.

Reply:
I asked both, extend `__builtin_add_overflow` for 128 bit, but typically there is a set like
`__builtin_[su]addl{0,2}_overflow` too, I think it would be good to keep that pattern from Clang's PoV. Having a new builtin also makes it easy for us to query whether `__builtin_add_overflow` has 128-bit support.

Reply:
`__builtin_add_overflow` already works with 128-bit integers: https://godbolt.org/z/non8eMTfj, as well as bit-precise integers: https://godbolt.org/z/dxs7h4G5b so I'm not certain there are Clang changes needed?

Reply:
You'll probably want to add test coverage for `_BitInt` as well given that it's an extension Clang supports in C++.

Reply:
> `__builtin_add_overflow` already works with 128-bit integers: https://godbolt.org/z/non8eMTfj, as well as bit-precise integers: https://godbolt.org/z/dxs7h4G5b so I'm not certain there are Clang changes needed?

Thanks @AaronBallman there are indeed no changes needed. I just need to learn how to read ;-)